src/ophys_etl/qc/video/correlation_graph.py

```diff
@@ -0,0 +1,152 @@
+
+    Parameters
+    ----------
+    video_path: Path
+        path to an hdf5 video file, assumed to have a dataset "data"
+        nframes x nrow x ncol
+    row_indices: List[int]
+        like [start, stop] and a dimension of the dataset will be directly
+        indexes like array[start: stop]
+    col_indices: List[int]
+        like [start, stop] and a dimension of the dataset will be directly
+        indexes like array[start: stop]
+
+    Returns
+    -------
+    graph: nx.Graph
+        an undirected networkx graph, with weighted edges
+
+    """
+
+    with h5py.File(video_path, "r") as f:
+        data = f["data"][:,
+                         row_indices[0]: row_indices[1],
+                         col_indices[0]: col_indices[1]]
+
+    # relative indices of 8 nearest-neighbors
+    nbrs = list(itertools.product([-1, 0, 1], repeat=2))
+    nbrs.pop(nbrs.index((0, 0)))
+
+    nrow = data.shape[1]
+    ncol = data.shape[2]
+    r0 = row_indices[0]
+    c0 = col_indices[0]
+
+    graph = nx.Graph()
+    for irow in range(nrow):
+        for icol in range(ncol):
+            edge_start = (irow, icol)
+            global_edge_start = (irow + r0, icol + c0)
+            edge_ends = []
+            global_edges = []
+            for dx, dy in nbrs:
+                # a local edge, appropriate for indexing the loaded data
+                edge_end = (irow + dy, icol + dx)
+                if (edge_end[0] < 0) | (edge_end[0] >= nrow):
+                    continue
+                if (edge_end[1] < 0) | (edge_end[1] >= ncol):
+                    continue
+                # a global edge, appropriate for the entire FOV
+                global_edge_end = (edge_end[0] + r0, edge_end[1] + c0)
+                edge_ends.append(edge_end)
+                global_edges.append([global_edge_start, global_edge_end])
+            # cdist 2.5x faster for 1-to-many than repeated calls to pearsonr
+            weights = 1.0 - cdist([data[:, edge_start[0], edge_start[1]]],
+                                  [data[:, edge_end[0], edge_end[1]]
+                                   for edge_end in edge_ends],
+                                  metric="correlation")[0]
+            for global_edge, weight in zip(global_edges, weights):
+                if not graph.has_edge(*global_edge):
+                    graph.add_edge(*global_edge, weight=weight)
+
+    return graph
+
+
+class CorrelationGraph(argschema.ArgSchemaParser):
+    default_schema = CorrelationGraphInputSchema
+
+    def run(self):
+        self.logger.name = type(self).__name__
+        t0 = time.time()
+
+        with h5py.File(self.args["video_path"], "r") as f:
+            nrow, ncol = f["data"].shape[1:]
+        row_indices = index_split(nrow, self.args["n_segments"])
+        col_indices = index_split(ncol, self.args["n_segments"])
+        args = [(self.args["video_path"], rows, cols)
+                for rows, cols in itertools.product(row_indices, col_indices)]
```

Comment:
No change requested, but you could possibly eliminate a function.

It looks like `index_split` and this `itertools.product` are trying to divide the 2D image into roughly equal sized sub 2D arrays for multiprocessing?

Maybe the `index_split` function could be eliminated altogether with something like:

```
h_split_indices = np.cumsum([len(subsplit) for subsplit in np.array_split(range(nrow), n_segments)[:-1]])
v_split_indices = np.cumsum([len(subsplit) for subsplit in np.array_split(range(ncol), n_segments)[:-1]])

[np.split(arr, v_split_indices, axis=1) for arr in np.split(img, h_split_indices, axis=0)]
```

`np.split` is supposed to return a view (https://numpy.org/doc/stable/reference/generated/numpy.split.html) so in theory should have similar memory implications as passing indices.

Reply:
I am only passing indices rather than the split video, so that each worker will separately open and read their designated part of the hdf5 dataset.
I am passing the index endpoints, rather than an indexing array, because:
```
d = f["data"][:, [0, 1], [0, 1]]
```
results in:
```
TypeError: Only one indexing vector or array is currently allowed for advanced selection

```
while:
```
d = f["data"][:, 0:2, 0:2]
```
is allowed.

And, the key feature is that the worker indices must overlap so that the subgraphs resulting from the mp workers are fully connected on `nx.compose_all()`. So, even though `index_split` is just `np.array_split` there is a tweak to enforce that overlap, and to make the `a:b` endpoint indices